File: lessons_follow_along_code/04_intro_differential_privacy.R
library(tidyverse)
library(urbnthemes)

# --- Setup ----

# package for laplace distribution
library(smoothmest)

# function to draw Laplace noise for one statistic 
laplace_mechanism <- function(sensitivity, epsilon){
  
  # lambda (distribution width) is sensitivity/privacy loss budget
  l <- sensitivity/epsilon 
  
  # draw from Laplace distribution
  noise <- rdoublex(n = 1, # draw one observation (adding noise to one statistic)
                    mu = 0, # centered at 0
                    lambda = l) # scale based on l calculated above
  
  return(noise)
}

# set seed for reproducibility
set.seed(26042022)

# calculate noise
laplace_mechanism(sensitivity = 8, epsilon = 1)

# calculate noise
laplace_mechanism(sensitivity = 1, epsilon = 8)

# --- Exercise 2 ---------

laplace_mechanism(sensitivity = #____,
                  epsilon = #____)

File: lessons_follow_along_code/04_intro_differential_privacy.py
import pandas as pd
import numpy as np

# --- Setup ---------

# Define function to add laplace noise
def laplace_mechanism(sensitivity, epsilon, n = 1):
  
  # lambda (distribution width) is sensitivity/privacy loss budget
  l = sensitivity / epsilon 
  
  result = np.random.laplace(scale = l, size = n)

  return(result)


# --- Exercise 2 ----------

# calculate noise
laplace_mechanism(sensitivity = #_____,
                  epsilon = #_____)

